Introducing new Action for closing gaps in tracks

haesleinhuepf opened the pull request:
Hey guys,

we wanted to do FRAP analysis. Therefore, it is neccessary to measure the signal of a spot even in frames, when it is not visible. The trackers allow building tracks with gaps, but measurement of the signal in these gaps was not possible. Thus, I introduced the action in this pull request allowing our analysis by introducing new spots in the track at empty frames by interpolating the position before and after the gap. Hopefully it is useful for others.

Cheers,
Robert

tinevez commented:
Great! It corresponds to several requests I had, also in lineaging.
I will test it very soon.

Looking at the core I could not ascertain that your strategy would work for tracks with cell divisions etc... Do you know this already?

tinevez commented:
Ok, I checked and it does not deal when you have e.g. a cell division. This is because you rely on the iterating over a collection of spots for each track.

You want to use a graph iterator, DepthFirstIterator, to traverse the track by following edges. 
I will add comments in the commit.

tinevez commented on src/main/java/fiji/plugin/trackmate/action/CloseGapsByLinearInterpolationAction.java:
```diff
@@ -0,0 +1,157 @@
+        // go through all tracks
+        for (Integer trackID : trackIDs)
+        {
+            Set<Spot> spots = model.getTrackModel().trackSpots(trackID);
```
Use TrackModel.getDepthFirstIterator instead.
Then test the action on a track that divides.

haesleinhuepf replied:
Hi Jean-Yves,

just a quick quesiton regarding this... 

If I had a dividing track like

```
A
|
B
|
C -- E
|
D
```

I get an iterator order like A,B,C,D,E

How do I know, that there is a division from C to E and C to D ?

tinevez replied:
You don't need to know. 
For this process, you just need to know whether two spots **that are connected via an edge** are separated by more than 1 frame.
The depthFistIterator just ensures that you iterate through a track by walking over edges. But of course, you need to maintain a source spot and a target spot as you iterate. 

Hey forget it actually, I have another idea. 
Just work on the **list of edges**. An edge as always a source and target spot, and these are the spots you operate on. You don't need to iterate over the track using the iterator I gave you.

haesleinhuepf replied:
I had the same idea :) look at my changes ;)

tinevez commented on src/main/java/fiji/plugin/trackmate/action/CloseGapsByLinearInterpolationAction.java:
```diff
@@ -0,0 +1,157 @@
+ * Date: June 2016
+ *
+ */
+public class CloseGapsByLinearInterpolationAction extends AbstractTMAction {
```
Could you reformat the code to comply to imglib2 style, using the imglib2 formatter if you use Eclipse?
It is there https://github.com/imglib/imglib2/blob/master/doc/imglib2-eclipse-mars-clean-up-formatter-style.xml

tinevez commented on src/main/java/fiji/plugin/trackmate/action/CloseGapsByLinearInterpolationAction.java:
```diff
@@ -0,0 +1,157 @@
+                            Spot newSpot = new Spot(rp, 0, 0);
+
+                            // Set some properties of the new spot
+                            interpolateFeature(newSpot, currentSpot, nextSpot, weight, "RADIUS");
```
Please use the constants defined in the analyzer of the Spot class instead of String literals. 
For example, use `Spot.QUALITY`, `Spot.RADIUS`, etc...

haesleinhuepf commented:
Thanks for the comments. I'm quite new in the imglib2 field and rely on feedback :) I will update the code asap.

tinevez commented:
Ping me if you need any help. This PR is important.

tinevez commented:
Also: Do you prefer shipping it as an independent module or in the core of TrackMate?

haesleinhuepf commented:
I've no preference. At the moment I have a single user using it. Let's see how/if this evolves. 

I could imagine that the action-pulldown becomes confusing if the number of entries increases and increases. How would the alternative to core of TrackMate work/look like?

tinevez commented:
A jar file that contains only your action. End users drop them in their Fiji plugins folder to get the functionality.

haesleinhuepf commented:
And I guess through the miracles of scijava annotations it is included in the pulldown as well, right?

tinevez commented:
Yep.

haesleinhuepf commented:
Cool technology. I could go this way for the moment. But then we can close the PR...

imagejan commented:
Hi guys,
this is a great addition, but if you go for shipping it separately:
> End users drop them in their Fiji plugins folder to get the functionality.

Please use an *update site* and don't let users mess around with their `plugins` folder!

I've seen things go wrong too often because people were advised to install this and that into their Fiji.app folders. While this is great for quickly adding and testing functionality, we should use the update site mechanism for "end users" whenever possible. ;)

haesleinhuepf commented:
Good point. I have some question about that, but as it is a bit off-topic, I create a thread in the forum: http://forum.imagej.net/t/how-to-deliver-plugins-to-users-update-site-mechanism/1925

haesleinhuepf commented:
@tinevez I updated the code according to your suggestions. The major changes are:
* I'm now going through all edges to check for gaps
* I added a test class showing that it works for linear and dividing tracks

If you have time, please check again. Thanks! :)

tinevez commented on src/main/java/fiji/plugin/trackmate/action/CloseGapsByLinearInterpolationAction.java:
```diff
@@ -0,0 +1,159 @@
+			for ( DefaultWeightedEdge edge : edges )
+			{
+				Spot currentSpot = trackModel.getEdgeSource( edge );
+				Spot nextSpot = trackModel.getEdgeTarget( edge );
```
Careful here: You assume that the target is always subsequent in time. 
This is true for "normal" TrackMate but I am always worry of weird implementation where the direction of the edge would be reversed or random. 
Could make it able to handle any weird case where time order would be reversed for source and target?

haesleinhuepf replied:
I can. :) See my recent changes. I also added a corresponding test case.

tinevez commented:
Checked manually: it works beautifully. Particularly useful with step-wise time browsing (F & G).
Thank you so much! This will be very useful to users. 

Now depending on what you want I can merge it to the core, or you can have it as an extra module. 
Tell me what you favor.

haesleinhuepf commented:
Merge it to the core. I would be happy to contribute something if it's useful :)

tinevez commented:
Welcome to the contributors.
